infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/server/ServersReadinessChecker.java

```diff
@@ -0,0 +1,175 @@
+                                   ServerCheckerFactory serverCheckerFactory) {
+        this.machineName = machineName;
+        this.servers = servers;
+        this.serverReadinessHandler = serverReadinessHandler;
+        this.serverCheckerFactory = serverCheckerFactory;
+        this.timer = new Timer("ServerReadinessChecker", true);
+    }
+
+    /**
+     * Asynchronously starts checking readiness of servers of a machine.
+     *
+     * @throws InternalInfrastructureException
+     *         if check of a server failed due to an unexpected error
+     * @throws InfrastructureException
+     *         if check of a server failed due to an error
+     */
+    public void startAsync() throws InfrastructureException {
+        List<ServerChecker> serverCheckers = getServerCheckers();
+        // should be completed with an exception if a server considered unavailable
+        CompletableFuture<Void> firstNonAvailable = new CompletableFuture<>();
+        CompletableFuture[] checkTasks = serverCheckers
+                .stream()
+                .map(ServerChecker::getReportCompFuture)
+                .map(compFut -> compFut.thenAccept(serverReadinessHandler)
+                                       .exceptionally(e -> {
+                                           // cleanup checkers tasks
+                                           timer.cancel();
+                                           firstNonAvailable.completeExceptionally(e);
+                                           return null;
+                                       }))
+                .toArray(CompletableFuture[]::new);
+        resultTimeoutSeconds = checkTasks.length * 180;
+        // should complete when all servers checks reported availability
+        CompletableFuture<Void> allAvailable = CompletableFuture.allOf(checkTasks);
+        // should complete when all servers are available or any server is unavailable
+        result = CompletableFuture.anyOf(allAvailable, firstNonAvailable);
+        for (ServerChecker serverChecker : serverCheckers) {
+            serverChecker.start();
+        }
+    }
+
+    /**
+     * Waits until servers are considered available or one of them is considered as unavailable.
+     *
+     * @throws InternalInfrastructureException
+     *         if check of a server failed due to an unexpected error
+     * @throws InfrastructureException
+     *         if check of a server failed due to interruption
+     * @throws InfrastructureException
+     *         if check of a server failed because it reached timeout
+     * @throws InfrastructureException
+     *         if check of a server failed due to an error
+     */
+    public void await() throws InfrastructureException {
+        try {
+            // TODO how much time should we check?
+            result.get(resultTimeoutSeconds, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
Have you considered propagating `InterruptedException` so `DockerRuntimeInfrastructure` decides what to do with it, i think it simplifies interruption policy as all the logic related to interruption processing placed on the same "plate"?